Add docs and short guide for gateways on k8s

lkysow opened the pull request:
Reference docs for the new helm values are still in progress.

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
```diff
@@ -0,0 +1,321 @@
+page_title: "Mesh Gateways - Kubernetes (beta)"
+sidebar_current: "docs-platform-k8s-mesh-gateways"
+description: |-
+    How to configure Connect Mesh Gateways in Kubernetes.
```
```suggestion
    This documentation describes how to configure Connect Mesh Gateways in Kubernetes.
```

lkysow commented:
@judithpatudith looks like the integration tests are failing which seems unrelated. I'm going to split this into two PRs: this one and another one with the reference docs for the config. This one is probably higher priority because it would be good to have something to point to for how to run gateways.

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
```diff
@@ -0,0 +1,360 @@
+We're going to use the load balancer's IP to route to the gateways so we need
+to wait for that to be up.
+
+Run
```
```suggestion
Search for the load balancer service. 
```

judithpatudith replied:
The relationship between the loadbalancer and the mesh gateway feels a bit squishy here. I think we're looking for . the loadbalancer that sits in front of the gateway, but it might be that the gateway is the loadbalancer.

lkysow replied:
Yeah it's a bit confusing because often a loadbalancer and a gateway are the same thing. In our case, the gateway is just an app we're running in Kubernetes and since other dc's need to route to it, we need a way to expose it externally... hence the load balancer. I think I can make it clearer though.

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+#  ...
+```
+
+### Step 2 - Configure Your First DC
````
```suggestion
### Step 2 - Configure Your First Datacenter
```

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+helm upgrade <your release name> . -f dc1-values.yaml
+```
+
+### Step 4 - Configure Your Second DC
````
```suggestion
### Step 4 - Configure Your Second Datacenter
```

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+```
+
+### Step 4 - Configure Your Second DC
+Now we're ready to start Consul in our second DC (`dc2`).
````
```suggestion
Now we're ready to start Consul in our second datacenter (`dc2`).
```

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+# You should also have the config specified in the prerequisites.
+```
+
+And then upgrade your chart:
````
```suggestion
And then upgrade your chart:

```

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
```diff
@@ -0,0 +1,360 @@
+We're going to use the load balancer's IP to route to the gateways so we need
+to wait for that to be up.
+
+Run
```
Same comments for the first datacenter's section apply here

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+helm upgrade <your release name> . -f dc2-values.yaml
+```
+
+### Step 6 - Ensure each datacenter can access the Load Balancers
````
Is this correct? I think the idea was to have the gateways talk to each other so that you didn't have to expose sidecar proxy pods to the outside world directly.

lkysow replied:
Yes you're totally correct! Good catch, I'll rework this :D

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+curl: (35) Unknown SSL protocol error in connection to 32.32.32.32:443
+```
+
+### Step 7 - Test it out
````
```suggestion
### Step 7 - Test it out

```

judithpatudith commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+helm upgrade <your release name> . -f dc2-values.yaml
+```
+
+### Step 6 - Ensure each datacenter can access the Load Balancers
````
```suggestion
### Step 6 - Ensure Each Datacenter Can Access the Load Balancers
```

judithpatudith commented:
Also I didn't look at the reference docs because I think they'll be separated out into another PR? Let me know if you'd like me to comment on those here too 🙂

lkysow commented:
If you don't mind reviewing them in this PR then go ahead. Otherwise I can
split it out.

On Thu, Jul 11, 2019, 6:45 PM Judith Malnick <notifications@github.com>
wrote:

> Also I didn't look at the reference docs because I think they'll be
> separated out into another PR? Let me know if you'd like me to comment on
> those here too 🙂
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hashicorp/consul/pull/6099?email_source=notifications&email_token=AAH4RPNIJRHUGWFRE7AWLO3P65WTHA5CNFSM4H7ISRZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZXOQPQ#issuecomment-510584894>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAH4RPLKFJYLB2FMFTNTSHTP65WTHANCNFSM4H7ISRZA>
> .
>

lkysow commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+          "primary_datacenter": "dc1"
+        }
+    ```
+1. If using Kubernetes, Connect injection must be enabled. This can be set via the helm chart:
````
I was thinking about the case of connecting a k8s and non-k8s datacenter together.

lkysow replied:
But maybe these prerequisites should be part of the main mesh gateway docs and then I can just reference them?

judithpatudith replied:
Yup. Looks like you captured most of these in the example helm charts, so I'm ok with adding the settings themselves to the Gateway docs and leaving them off here. The one that we should keep is the note about Wan joined datacenters, but I don't think you do that till both your DCs are up and running, so its actually part of the process and not a prerequisite

lkysow commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+       enabled: true
+    ```
+
+## Guide to Using Mesh Gateways To Connect Two Kubernetes Clusters
````
Yes I do. I think its eventual home should be there, but I'm worried that it's not complete enough to live there in its current state. I think having at least these docs will be useful until we get a proper guide together but am happy to hear otherwise.

lkysow commented on website/source/docs/platform/k8s/mesh-gateways.html.md:
````diff
@@ -0,0 +1,360 @@
+#  ...
+```
+
+### Step 2 - Configure Your First DC
````
It will only make changes to what's different. So if I haven't changed any of the, say, client config, then they won't be affected. Depending on what is changed, user intervention might be required. For example, changing the extraConfig section requires manually restarting or reloading each component.

lkysow commented:
@judithpatudith I've addressed your feedback as follows:
* Split out code blocks, added description in text above, added `$` and added output where applicable
* Removed the section on ensuring Pods can talk to the load balancers 🍻 
* Move prerequisite YAML values into the values.yaml in the steps as well
* Added a part to the guide that demonstrates how we're unable to route between the services without the gateway

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+* <a name="v-meshgateway" href="#v-meshgateway">`meshGateway` <sup>(beta)</sup></a> - Values that configure Connect [Mesh Gateways](/docs/connect/mesh_gateway.html).
+
+  * <a name="v-meshgateway-enabled" href="#v-meshgateway-enabled">`enabled`</a> (`boolean: false`) - If true, gateways will be created and
+    Consul servers will be configured to support them. If setting this in an already running cluster, you'll need to
```
This sounds a bit like setting enabled to true will take care of the prereqs. Might be clearer if we put this sentence in the active voice. Also, I'm not 100% sure about what exactly "gateways will be created" means. Does this schedule, configure, and manage envoy?

lkysow replied:
What about:
```
Setting to true will cause mesh gateways to be deployed.
```

lkysow replied:
> Does this schedule, configure, and manage envoy?

Yes it creates a Deployment with Pods that run envoy and register as gateways.

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  If set to the empty string, no global default will be set and the gateway mode
+  will need to be set individually for each service.
+
+  * <a name="v-meshgateway-replicas" href="#v-meshgateway-replicas">`replicas`</a> (`int: 2`) -
```
```suggestion
  * <a name="v-meshgateway-replicas" href="#v-meshgateway-replicas">`replicas`</a> (`int: 2`) -
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  will need to be set individually for each service.
+
+  * <a name="v-meshgateway-replicas" href="#v-meshgateway-replicas">`replicas`</a> (`int: 2`) -
+  How many gateway Pods to run.
```
```suggestion
  How many gateway pods to run.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  How many gateway Pods to run.
+
+  * <a name="v-meshgateway-wanaddress" href="#v-meshgateway-wanaddress">`wanAddress`</a> -
+  Values that configure what gets registered as the wan address for the gateway when running the `consul connect envoy` command.
```
```suggestion
  The nested values define the WAN address registration that the `consul connect envoy` command will use to configure the mesh gateway.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  See the [command overview](/docs/commands/connect/envoy.html#wan-address) for more details.
+
+      * <a name="v-meshgateway-wanaddress-port" href="#v-meshgateway-wanaddress-port">`port`</a> (`int: 443`) -
+      The port that gets registered for the wan address.
```
```suggestion
      The port that gets registered for the WAN address.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+
+      * <a name="v-meshgateway-wanaddress-usenodeip" href="#v-meshgateway-wanaddress-usenodeip">`useNodeIP`</a> (`bool: true`) -
+      If true, each Gateway Pod will advertise its Node IP
+      (as provided by the Kubernetes downward API) as the wan address.
```
```suggestion
      (as provided by the Kubernetes downward API) as the WAN address.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+      If true, each Gateway Pod will advertise its Node IP
+      (as provided by the Kubernetes downward API) as the wan address.
+      This is useful if node IPs are routable from other DCs.
+      `useNodeName` and `host` must be `false` and `""` respectively.
```
```suggestion
    If you use the node IP, then `useNodeName` and `host` must be `false` and `""` respectively.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+      If true, each Gateway Pod will advertise its Node Name
+      (as provided by the Kubernetes downward API) as the wan address.
+      This is useful if the node names are DNS entries that are routable from other DCs.
+      `useNodeIP` and `host` must be `false` and `""` respectively.
```
```suggestion
     If you use the node name, then `useNodeIP` and `host` must be `false` and `""` respectively.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+
+      * <a name="v-meshgateway-wanaddress-usenodename" href="#v-meshgateway-wanaddress-usenodename">`useNodeName`</a> (`bool: false`) -
+      If true, each Gateway Pod will advertise its Node Name
+      (as provided by the Kubernetes downward API) as the wan address.
```
I think WAN should be capitalized everywhere? 🤔

judithpatudith replied:
Except in the actual settings of course

lkysow replied:
Yes I think you're right, I'll run through and change it.

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  will need to be set individually for each service.
+
+  * <a name="v-meshgateway-replicas" href="#v-meshgateway-replicas">`replicas`</a> (`int: 2`) -
+  How many gateway pods to run.
```
```suggestion
  How many gateway Pods to run.
```

judithpatudith replied:
Sorry I thought that "Pod" would be lowercase but it looks like they capitalize it in the K8s docs so I guess we should do the same.

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+      `useNodeIP` and `host` must be `false` and `""` respectively.
+
+      * <a name="v-meshgateway-wanaddress-host" href="#v-meshgateway-wanaddress-host">`host`</a> (`string: """`) -
+      If non-empty, each gateway Pod will use this host as its wan address.
```
```suggestion
      If non-empty, each gateway Pod will use this host as its wan address.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+
+      * <a name="v-meshgateway-wanaddress-host" href="#v-meshgateway-wanaddress-host">`host`</a> (`string: """`) -
+      If non-empty, each gateway Pod will use this host as its wan address.
+      Users must ensure that this address routes to the Gateway pods,
```
```suggestion
      Users must ensure that this address routes to the Gateway Pods,
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+      for example via a load balancer IP or a DNS entry that routes to the Service
+      fronting the Deployment.
+      `meshGateway.wanAddress.port` will be used as the port for the wan address.
+      `useNodeIP` and `useNodeName` must be false.
```
```suggestion
      If you specify a host, then `useNodeIP` and `useNodeName` must be false.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+      `useNodeIP` and `useNodeName` must be false.
+
+  * <a name="v-meshgateway-service" href="#v-meshgateway-service">`service`</a> -
+  Values that configure the Kubernetes `Service` for the gateway `Deployment.
```
```suggestion
  The nested values configure the Kubernetes `Service` for the gateway `Deployment.
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  Values that configure the Kubernetes `Service` for the gateway `Deployment.
+
+      * <a name="v-meshgateway-service-enabled" href="#v-meshgateway-service-enabled">`enabled`</a> (`bool: false`) -
+      Whether to create a Service or not.
```
This might be my own ignorance about K8s, but if creating a service for the gateways is optional it might be good to explain why you would or wouldn't do it.

lkysow replied:
👍 will do

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
````diff
@@ -377,6 +377,176 @@ to run the sync program.
+          clusterIP: 10.0.171.239
+        ```
+
+  * <a name="v-meshgateway-imageenvoy" href="#v-meshgateway-imageenvoy">`imageEnvoy`</a> (`string: "envoyproxy/envoy:v1.10.0"`) -
````
For my own info: We aren't providing a way for users to pick their Gateway proxy (like we do for sidecars)? Is this on the roadmap?

lkysow replied:
Only envoy is supported right now. Not sure if other proxies are on the roadmap.

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+  will fail health checks. You may disable health checks as a temporary
+  workaround.
+
+  * <a name="v-meshgateway-resources" href="#v-meshgateway-resources">`resources`</a> (`string`) -
```
All the code blocks under bullet points on this page look like they might need to be indented one more level to nest correctly under the bullets.

lkysow replied:
👍

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
````diff
@@ -377,6 +377,176 @@ to run the sync program.
         ```
+* <a name="v-meshgateway" href="#v-meshgateway">`meshGateway` <sup>(beta)</sup></a> - Values that configure Connect [Mesh Gateways](/docs/connect/mesh_gateway.html).
+
+  * <a name="v-meshgateway-enabled" href="#v-meshgateway-enabled">`enabled`</a> (`boolean: false`) - If true, gateways will be created and
````
```suggestion
  * <a name="v-meshgateway-enabled" href="#v-meshgateway-enabled">`enabled`</a> (`boolean: false`) - If true, mesh gateways will be enabled on the Consul servers, registered with Consul according to the nested parameters below, and started by Kubernetes. 
```

judithpatudith commented on website/source/docs/platform/k8s/helm.html.md:
```diff
@@ -377,6 +377,176 @@ to run the sync program.
+* <a name="v-meshgateway" href="#v-meshgateway">`meshGateway` <sup>(beta)</sup></a> - Values that configure Connect [Mesh Gateways](/docs/connect/mesh_gateway.html).
+
+  * <a name="v-meshgateway-enabled" href="#v-meshgateway-enabled">`enabled`</a> (`boolean: false`) - If true, gateways will be created and
+    Consul servers will be configured to support them. If setting this in an already running cluster, you'll need to
```
```suggestion
    CIf setting this in an already running cluster, you'll need to
```

judithpatudith commented:
@lkysow I think I'm done looking through the settings doc too now. Let me know if you have any questions. I'm gonna move those prereqs to the gateway doc

judithpatudith commented:
The PR for the prerequisites is #6126

lkysow commented:
I'm going to hold off on this one until we get server wan joining in place.

i0rek commented:
Is this still relevant?

lkysow commented:
> Is this still relevant?

It can be closed for now. When we add mesh federation then we'll probably re-use some of this content.
